library/alloc/src/rc.rs

```diff
@@ -2256,17 +2256,14 @@ unsafe impl<#[may_dangle] T: ?Sized, A: Allocator> Drop for Rc<T, A> {
         unsafe {
             self.inner().dec_strong();
             if self.inner().strong() == 0 {
-                // destroy the contained object
-                ptr::drop_in_place(Self::get_mut_unchecked(self));
-
-                // remove the implicit "strong weak" pointer now that we've
-                // destroyed the contents.
-                self.inner().dec_weak();
-
-                if self.inner().weak() == 0 {
-                    self.alloc
-                        .deallocate(self.ptr.cast(), Layout::for_value_raw(self.ptr.as_ptr()));
-                }
+                // Reconstruct the "strong weak" pointer and drop it when this
+                // variable goes out of scope. This ensures that the memory is
+                // deallocated even if the destructor of `T` panics.
+                let _weak = Weak { ptr: self.ptr, alloc: &self.alloc };
+
+                // Destroy the contained object.
+                // We cannot use `get_mut_unchecked` here, because `self.alloc` is borrowed.
+                ptr::drop_in_place(&mut (*self.ptr.as_ptr()).value);
```

Comment:
Could this just use raw mut?

```rust
ptr::drop_in_place(&raw mut self.ptr.as_ptr().value);
```

Reply:
As far as I understand, `drop_in_place` always works as if it's argument was `&mut T` and only uses `*mut T` in the signature for compatibility reasons (see https://github.com/rust-lang/rust/pull/103957 and https://github.com/rust-lang/rust/pull/111807).

So, yes, this could use `&raw mut`, but the pointer would just automatically be reborrowed as a reference anyway, so i used `&mut` directly for clarity.

Reply:
Makes sense, thanks for the follow up